### PR TITLE
Retarget depending pulls when the parent branch is deleted

### LINKED ISSUE #18408: Allow cascade PRs
### Feature Description

Sometime you need to have 'cascade' PRs. i.e.

* first you're working on feature A (derived from main branch) and you create PR feature A -> main
* next you need to work on feature B, but this feature depends on changes of feature A and it's corresponding PR is not yet approved. The best way to work on such feature is derive from feature A and later you need to create PR feature B -> feature A.

Currently second PR will be automatically closed on feature A merge and branch delete. It would be helpfull to rebase second PR. i.e. change 'feature B -> feature A' to 'feature B -> main' when first 'feature A -> main' is merged and 'feature A' branch is deleted.

### Screenshots

_No response_

### LINKED ISSUE #27062: Changing the destination branch of a pull request after deletion
### Feature Description

`branch3 -> branch2 -> branch1 -> master`
Now, if we have a chain of pull requests, we cannot merge `branch1` first, because all other requests to `branch1` will be closed without the possibility of reopening.

In bitbucket, this was solved by the fact that when deleting a `branch1`, all open pull requests change the target to the branch into which `branch1` was merged, in our case, this is the master.

### Screenshots

Here is an example in the form of a mind map.
Schema before merge `branch1`:
Before merge
After merge and deleting `branch1`:
After merge

### PR DESCRIPTION
**Pull request by kvaster**
Sometimes you need to work on a feature which depends on another (unmerged) feature.
In this case, you may create a PR based on that feature instead of the main branch.
Currently, such PRs will be closed without the possibility to reopen in case the parent feature is merged and its branch is deleted.
Automatic target branch change make life a lot easier in such cases.
Github and Bitbucket behave in such way.

Example:
$PR_1$: main <- feature1
$PR_2$: feature1 <- feature2

Currently, merging $PR_1$ and deleting its branch leads to $PR_2$ being closed without the possibility to reopen.
This is both annoying and loses the review history when you open a new PR.

With this change, $PR_2$ will change its target branch to main ($PR_2$: main <- feature2) after $PR_1$ has been merged and its branch has been deleted.

This behavior is enabled by default but can be disabled.
For security reasons, this target branch change will not be executed when merging PRs targeting another repo. 

Fixes #27062
Fixes #18408

### CODE REVIEW

**Comment by kvaster** [comment]
@denyskon , my advice to review this PR.

And if you read previous thread:
* I was against feature flag - it can be removed, cause such behaviour seems to be normal for any case
* I'm not able to write proper unit tests right now as I was requested
* Child PR's are not retargeted in case Base PR targets another repo

File: custom/conf/app.example.ini
Comment (denyskon) [inline custom/conf/app.example.ini]:
```suggestion
;RETARGET_CHILDREN_ON_MERGE = true
```

File: docs/content/administration/config-cheat-sheet.en-us.md
Comment (denyskon) [inline docs/content/administration/config-cheat-sheet.en-us.md]:
```suggestion
- `RETARGET_CHILDREN_ON_MERGE`: **true**: Retarget child pull requests to the parent pull request branch target on merge of parent pull request.
```

File: modules/setting/repository.go
Comment (denyskon) [inline modules/setting/repository.go]:
```suggestion
			RetargetChildrenOnMerge:                  true,
```

**Review by denyskon — APPROVED** [review]
I am not against making this feature optional, but definitely enabled by default. I think it is good that retarget only occurs inside the same repo, because otherwise PRs may end up somewhere they weren't supposed to.

**Comment by denyskon** [comment]
@kvaster If we need a test, you can add me as collaborator to your fork, then I'd try to write one.

**Comment by denyskon** [comment]
@kvaster Could you mention closing #27062 in PR description? It has a better description of the feature, IMO

**Comment by kvaster** [comment]
> I am not against making this feature optional, but definitely enabled by default. I think it is good that retarget only occurs inside the same repo, because otherwise PRs may end up somewhere they weren't supposed to.

I think that making this as a feature is non usefull at all. I will remove with please feature gate if gitea authors/maintainers are ok with this. So, waiting for agreement on this.

@denyskon, adding you as a collaborator.

**Comment by denyskon** [comment]
It's normal for Gitea to be able to disable everything you want, so I don't see any reason not to allow it here too 😆 It is already implemented, so let's keep it, but enable by default.

I'll try to take care of the test later today

**Comment by kvaster** [comment]
Great. Will update PR in a few minutes.

File: services/pull/pull.go
Comment (delvh) [inline services/pull/pull.go]:
```suggestion
// RetargetBranchPulls change target branch for all pull requests whose base branch is the branch
// Both branch and targetBranch must be in the same repo (for security reasons) 
```

File: services/pull/pull.go
Comment (delvh) [inline services/pull/pull.go:559]: Doesn't this method belong semantically into `models`?

**Review by delvh — APPROVED** [review]

File: services/pull/pull.go
Comment (ksa-real) [reply services/pull/pull.go:559]: Why so? IMO this is a helper method that should live nearby e.g. CloseBranchPulls

Comment (delvh) [reply services/pull/pull.go:559]: Well… Its only purpose is to update the DB.
That's exactly what the `models` package is supposed to do: Interact with the DB.
The `services` package is mainly intended for the actual logic.

File: docs/content/administration/config-cheat-sheet.en-us.md
Comment (delvh) [inline docs/content/administration/config-cheat-sheet.en-us.md]:
```suggestion
- `RETARGET_CHILDREN_ON_MERGE`: **true**: Retarget child pull requests to the parent pull request branch target on merge of parent pull request. It only works on merged PRs where the head and base branch target the same repo.
```

File: custom/conf/app.example.ini
Comment (delvh) [inline custom/conf/app.example.ini]:
```suggestion
;; Retarget child pull requests to the parent pull request branch target on merge of parent pull request. It only works on merged PRs where the head and base branch target the same repo.
```

**Review by delvh — APPROVED** [review]
I've finished testing, everything behaves exactly as expected :tada:

File: tests/integration/pull_merge_test.go
Comment (delvh) [inline tests/integration/pull_merge_test.go:472]: I think we should also add the second test with the following condition:
1. another user forks repo A
2. this user creates PR B targeting A
3. they also create PR C targeting B
4. B is merged
5. C is closed as a result of `4.`

Comment (denyskon) [reply tests/integration/pull_merge_test.go:472]: Done :)

File: services/pull/pull.go
Comment (denyskon) [reply services/pull/pull.go:559]: @delvh I tried, this isn't possible because of an import cycle issue when importing "models" inside "models/pull". I also don't see a reason to do so, honestly....

**Review by denyskon — APPROVED** [review]

File: services/pull/pull.go
Comment (kvaster) [reply services/pull/pull.go:559]: @delvh, the logic is a little bit more complex when just updating db. `ChangeTargetBranch` is used for retargeting and it is (and was) located exactly in ths same file.

**Comment by MGChecker** [comment]
Maybe one should take a look how retargeting like this affects CI integrations like Drone, in particular. I remember having observed issues regarding that in the past.

**Comment by delvh** [comment]
What do you mean?
What sort of issues?

**Comment by 6543** [comment]
If there is a problem with integration with other software, just open a new issue and we see how we can address it on our side.

For your specific note, tbh. that sounds more like a bug in drone-ci ... but if it realy ocures, let us know :)

**Pull request merged**
